### ISSUE TITLE
Fix crash when CUDA callbacks are executed during thread termination

### ISSUE THREAD

**Issue by DarkPlayer** [wine-compholio/wine-staging#258]
This is a split off from #231 

CUDA allows to add stream callbacks (`cuStreamAddCallback`) that are guaranteed to be executed once. Stream callbacks are used to notify the application about the end of a asynchronous memory transfer. In order to ensure that the callback is executed at least once (even if the thread terminates), CUDA creates a pthread thread locale variable using `pthread_key_create` with a destructor. This destructor is called by pthread as last step when a thread exits, for example during `pthread_exit`.

Wine does not expect any code to be execute after `pthread_exit`. Excerpt from `dlls/ntdll/thread.c`:
```c
void terminate_thread( int status )
{
    pthread_sigmask( SIG_BLOCK, &server_block_set, NULL );
    if (interlocked_xchg_add( &nb_threads, -1 ) <= 1) _exit( status );

    close( ntdll_get_thread_data()->wait_fd[0] );
    close( ntdll_get_thread_data()->wait_fd[1] );
    close( ntdll_get_thread_data()->reply_fd );
    close( ntdll_get_thread_data()->request_fd );
    pthread_exit( UIntToPtr(status) );
}
```
Wine closes the connection to the wineserver and expects the thread to terminate. If the stream callback was not executed yet, CUDA will now execute the function specified by the application during `pthread_exit`. If the callback tries to use any Windows API function which requires a wineserver connection, the wineserver call is failing causing the application to exit with the following error:
```
wine client error:3e: Bad file descriptor
```
One solution would be to defer the release of the wineserver connection but pthread does not seem to offer any way to add a cleanup callback after the release of the thread local storage. We will need to find a different way to workaround this problem.

**Comment by slackner** [comment]
Marking as bug because this can be seen as a flaw in our wrapper. Setting milestole to 1.7.35 because it would be nice to have a fix for that as soon as possible. ;)

**Comment by slackner** [comment]
Can be reproduced with the simpleCallback.exe demo from http://developer.download.nvidia.com/compute/cuda/6_0/rel/installers/cuda_6.0.37_winxp_general_32.exe

**Comment by slackner** [comment]
This might be more difficult than expected.

First of all, the idea of hooking pthread is bad, because the linux library loader itself uses pthread functions. Depending on how exactly the hook works it will break wine completely.

I then decided to do various tests by manually renaming import symbols with an hex editor. Even when those symbols are renamed the library still loads fine here, and the crash is identical to before. This probably means that our analysis is wrong, and that a different API function is used to setup thread exit handlers.

If we don't have any idea how to fix it properly it might be better to remove the milestone from this bug.

**Comment by slackner** [comment]
I think I figured out what exactly goes wrong. My initial assumption was wrong, the issue has to be solved differently. Good thing: We will probably not need any pthread hooks.

**Comment by slackner** [comment]
Fixed with fbe41916cb50c584a744412c2845f44da81ae1af.

Just for reference: The problem is that CUDA creates a new thread to execute stream callbacks. Since %fs is duplicated into the new thread, they both try to use the same wineserver connection, which was causing the trouble. By forwarding all callbacks to a separate worker thread this problem should be solved. Please test!

**Comment by slackner** [comment]
Closing as fixed.
